### PR TITLE
Add support for "Move on Open"

### PR DESCRIPTION
**Pull request by RosemaryOrchard**
Related to #830  and #1119, this adds support for "move on open". It's implemented as a checkbox below the "Default consume location" select, and is disabled that matches the default product location. If checked, when you mark a product as open it moves it from wherever it is, to the default consume location (and falls back to wherever the product currently is, just in case there is ever a scenario where the consume location is not set). 

This is my first PR to Grocy, and my first time working with Slim, so I hope I've not done anything too crazy—feedback is very welcome! If this goes well I'd like to add support for fridges next.

### CODE REVIEW

File: services/StockService.php [hunk 1]
```diff
@@ -1048,15 +1048,20 @@ public function OpenProduct(int $productId, float $amount, $specificStockEntryId
...
-				]);
+				];
 
 				$amount = 0;
 			}
+			if ($product->move_on_open) {
+				$update['location_id'] = $product->default_consume_location_id ?? $stockEntry->location_id;
+			}
+
+			$stockEntry->update($update);
```
Comment (berrnd) [inline services/StockService.php:1064]: To keep it consistent, transfer from/to stock journal entries should also be created when the product is "moved on open" (just like when the transfer transaction would have been done manually).

**Review by berrnd — CHANGES_REQUESTED** [review]
Thanks!

This is not what #830 describes exactly, but maybe is enough, also to keep it simple - thanks for the idea.

Don't know how common it is to have more then one fridge while also storing _the same_ product sometimes in fridge1 and the other day in fridge2 after opening it. So I would say that's enough.

File: public/viewjs/consume.js [hunk 2]
```diff
@@ -386,26 +388,22 @@ Grocy.Components.ProductPicker.GetPicker().on('change', function(e)
...
+								value: stockLocation.location_id,
+								text: stockLocationName
+							}));
+							$("#location_id").val(defaultLocationId);
+							$("#location_id").trigger('change');
+							setDefault = 1;
+							stockAmountAtDefaultLocation += Number.parseFloat(stockLocation.amount);
 
-							if (setDefault == 0)
+							if (!setDefault)
```
[the remark this answers is not in the document]
Comment (RosemaryOrchard) [reply public/viewjs/consume.js:406]: Oh whoops, this snuck in as I started with the fridge and thought I only staged the right parts, I'll revert that change!

File: public/viewjs/productform.js [hunk 3]
```diff
@@ -239,7 +239,24 @@ $('#product-form input').keyup(function(event)
...
+	var consumeLocationLocation = document.getElementById('default_consume_location_id');
+
+	var moveOnOpen = document.getElementById('move_on_open');
+	if (consumeLocationLocation && defaultLocation !== consumeLocationLocation) {
+		moveOnOpen.removeAttribute('checked');
+		moveOnOpen.setAttribute('disabled', "");
+	} else {
+		moveOnOpen.removeAttribute('disabled');
+	}
+}
```
[the remark this answers is not in the document]
Comment (RosemaryOrchard) [reply public/viewjs/productform.js]: Thanks!

File: services/StockService.php (hunk 1 above)
Comment (RosemaryOrchard) [reply services/StockService.php:1064]: Thinking about it, should I flip the logic and check first, and then set the `stock_log` straight to the new location if there is one?

Comment (RosemaryOrchard) [reply services/StockService.php:1064]: Read some more code: nope!

Comment (berrnd) [reply services/StockService.php:1064]: A normal/manual transfer action creates two stock journal entries, the negative amount at the "from location" and the positive amount at the "to location":

grafik

Is done by [`StockService->TransferProduct`](https://github.com/grocy/grocy/blob/0152f1c69d3cb2bb3ceba668aa1006caa19a91c6/services/StockService.php#L1180) maybe just call that function there...

Comment (RosemaryOrchard) [reply services/StockService.php]: Thanks for the pointer, it makes much more sense to use the existing function.

**Comment by RosemaryOrchard** [comment]
I would like to add support more in line with specifically what 830 was requesting, but this seemed like a less intrusive start—and solves the problem that I myself have right now.

**Comment by berrnd** [comment]
Yes, I would also say that's what practically makes the most sense. When #830 was requested, the "Default consume location" idea wasn't there yet (ref #1365) - so in combination with that it definitely fits best.

Thanks again, I'll test that again later or tomorrow.

**Comment by RosemaryOrchard** [comment]
Thank you very much! Both for the review, and the great app 😁

**Review by berrnd — APPROVED** [review]
👍, works.

I've additionally added some UI feedback when auto-moving happened to make that more transparent.

**Pull request merged**